Fix most Unicode encoding bugs

fmeum opened the pull request:
Bazel aims to support arbitrary file system path encodings (even raw byte sequences) by attempting to force the JVM to use a Latin-1 locale for OS interactions. As a result, Bazel internally encodes `String`s as raw byte arrays with a Latin-1 coder and no encoding information. Whenever it interacts with encoding-aware APIs, this may require a reencoding of the `String` contents, depending on the OS and availability of a Latin-1 locale.

This PR introduces the concepts of *internal*, *Unicode*, and *platform* strings and adds dedicated optimized functions for converting between these three types (see the class comment on the new `StringEncoding` helper class for details). These functions are then used to standardize and fix conversion throughout the code base. As a result, a number of new end-to-end integration tests for the handling of Unicode in file paths, command-line arguments and environment variables now pass. 

Full support for Unicode beyond the current active code page on Windows is left to a follow-up PR as it may require patching the embedded JDK.

* Replace ad-hoc conversion logic with the new consistent set of helper functions.
* Make more parts of the Bazel client's Windows implementation Unicode-aware. This also fixes the behavior of `SetEnv` on Windows, which previously would remove an environment variable if passed an empty value for it, which doesn't match the Unix behavior.
* Drop the `charset` parameter from all methods related to parameter files. The `ISO-8859-1` vs. `UTF-8` choice was flawed since Bazel's internal string representation doesn't maintain any encoding information - `ISO-8859-1` just meant "write out raw bytes", which is the only choice that matches what arguments would look like if passed on the command line.
* Convert server args to the internal string representation. The arguments for requests to the server were already converted to Bazel's internal string representation, which resulted in a mismatch between `--client_cwd` and `--workspace_directory` if the workspace path contains non-ASCII characters.
* Read the downloader config using Bazel's filesystem implementation.
* Make `MacOSXFsEventsDiffAwareness` UTF-8 aware. It previously used the `GetStringUTF` JNI method, which, despite its name, doesn't return the UTF-8 representation of a string, but modified CESU-8 (nobody ever wants this).
* Correctly reencode path strings for `LocalDiffAwareness`.
* Correctly reencode the value of `user.dir`.
* Correctly turn `ExecRequest` fields into strings for `ProcessBuilder` for `bazel --batch run`. This makes it possible to reenable the `test_consistent_command_line_encoding` test, fixing #1775.
* Fix encoding issues in `TargetCompleteEvents`.
* Fix encoding issues in `SubprocessFactory` implementations.
* Drop obsolete warning if `file.encoding` doesn't equal `ISO-8859-1` as file names are encoded with `sun.jnu.encoding` now.
* Consistently reencode internal strings passed into and out of `FileSystem` implementations, e.g. if reading a symlink target. Tests are added that verify the interaction between `FileSystem` implementations and the Java (N)IO APIs on Unicode file paths.

Fixes #1775 
Fixes #11602
Fixes #18293 
Work towards #374
Work towards #23859

fmeum commented:
Switching back to draft as there are some unexpected failures.

fmeum commented:
@meteorcloudy Could you review the changes to the Bazel client?
@meisterT Could you review the other changes? Feel free to suggest other reviewers, this change really cuts through the whole code base.

Sorry for the large PR, but I needed to make sure encodings are consistent everywhere to get the tests to pass. Without them, it would be very hard to tell whether we regressed any use case that happened to work before this change.

meisterT reviewed (commented):
I will hand this review over to @lberki

fmeum commented on .bazelci/presubmit.yml:
```diff
@@ -64,6 +64,9 @@ tasks:
     shell_commands:
       - rm -rf $HOME/bazeltest
       - mkdir $HOME/bazeltest
+      - sudo apt update
```
I talked to @meteorcloudy about this and we arrived at the conclusion that adding the locales to the image could cause churn for other projects running on Bazel CI.

meteorcloudy replied:
@fmeum Do you know any specific projects will be broken if we do this by default? I'm not very familiar with the encoding stuff so it's hard for me to judge. But sure we can also just test this and revert back to older image if it's too intrusive.

fmeum replied:
I removed these lines now that Yun pushed updated images.

lberki commented on src/main/java/com/google/devtools/build/lib/util/StringEncoding.java:
```diff
@@ -0,0 +1,169 @@
+package com.google.devtools.build.lib.util;
```
nit: add license blurb

lberki commented on src/main/java/com/google/devtools/build/lib/util/StringEncoding.java:
```diff
@@ -0,0 +1,169 @@
+ * <p>Bazel achieves this by forcing an en_US.ISO-8859-1 locale on Unix when available, which due to
+ * the byte-based nature of Unix APIs allows all Java (N)IO functions to treat strings as raw byte
+ * sequences (a Latin-1 character is equivalent to an unconstrained byte value). On macOS, where the
+ * JVM forces UTF-8 encoding for any kind of system interaction, as well as on Windows, where system
```
Woah. Is that so?

lberki commented on src/test/shell/integration/watchfs_test.sh:
```diff
@@ -99,4 +99,49 @@ function test_both() {
+
+genrule(
+    name = "foo",
+    srcs = ["foo 🌱.in"],
```
This doesn't work at the HEAD? According to my occasional poking on Linux, it does, but I didn't check any other operating systems.

lberki replied:
Nevermind. This is in `watchfs_test`.

fmeum commented on src/main/cpp/option_processor.cc:
```diff
@@ -635,6 +640,28 @@ static void PreprocessEnvString(const string* env_str) {
 }
 #endif  // defined(_WIN32)
 
+#ifdef _WIN32
```
Done!

fmeum reviewed (commented):
> I added some comments, but man, is this change huge. Can you split it up somehow so that I don't forget the first file I looked at before I reach the third of this change? (I gave reviewing this change a honest attempt but it made my head hurt...)

The problem is that I can't guarantee that the split off commits won't cause regressions as the new tests require essentially all changes in this PR to pass. I will try to find a safe subset that I can split.

fmeum commented:
@lberki I split the PR up into five individual commits that can all be reviewed in isolation. The fourth one already adds some of the new tests. Unfortunately GitHub ain't no Critique and stacked PRs are pretty horrible - is the split into commits good enough for a decent review experience or should I split it into five individual PRs?

lberki commented:
There are two, mostly equally valid ways to make things right: one can either do what you are doing in this change and commit to the "bytes as Latin-1" approach or dispense with it, use UTF-8 everywhere and not support files with badly encoded names in the future.

I did some archaeology and if this bug I found from 2008 February (not a typo) is any indication, the reason why the "bytes as Latin-1" approach was taken is indeed file names with bad encoding. I suspect that now that UTF-8 is way more prevalent, that's much less of a problem today. 

I don't have a strong opinion (currently, this might change the more I think about it) about which way Bazel should go other than that "bytes as Latin-1" is definitely the odd choice. Have you considered this approach?

fmeum commented on src/main/cpp/main.cc:
```diff
@@ -32,3 +33,21 @@ int main(int argc, char **argv) {
+  for (const std::string &arg : args) {
+    c_args.push_back(const_cast<char *>(arg.c_str()));
+  }
+  return main_impl(c_args.size(), c_args.data());
```
Thanks, done

fmeum commented:
> I don't have a strong opinion (currently, this might change the more I think about it) about which way Bazel should go other than that "bytes as Latin-1" is definitely the odd choice. Have you considered this approach?

Yes, and I think that we should ultimately switch to it to reduce the complexity of the code. But I think that getting "bytes as Latin-1" done right is the safer path to that goal as it allows us to introduce meaningful end-to-end tests in a backwards compatible way. Once we are at that point, we can reason about what use cases we would break if we transitioned to the "everything UTF-8" approach. If we do that right away, we may end up breaking users in unforeseen ways.

Roughly half of the changes in this PR are needed for both approaches anyway. If we want to fully migrate to UTF-8 later, we essentially only need to start using the new encoding functions in `UnixFileSystem`, inline them to no-ops in all other places and consume files as UTF-8 rather than Latin-1. I would be happy to help with this and suggest changes, but this is ultimately something a Googler would be better positioned to do, simply because of the internal-only Linux distributions being used at Google that I have no knowledge of (and that, for example, don't yet have a `C.UTF-8` locale).

lberki commented:
Yeah, I agree that if we go with the "use proper UTF-8" approach, Googlers are in a better position to do so. In fact, there is a change from a few years ago floating about that just did that; I don't remember what its fate was.

Let's see what @tjgq and @coeuvre and @justinhorvitz think! I'm inclined to accept this pull request for precisely the reasons you mention: it fixes the interface of Bazel and leaves us free to refactor the internals, should we end up deciding that that's the right thing to do but I don't want to decide all by myself.

tjgq commented:
> Yeah, I agree that if we go with the "use proper UTF-8" approach, Googlers are in a better position to do so. In fact, there is a change from a few years ago floating about that just did that; I don't remember what its fate was.
> 
> Let's see what @tjgq and @coeuvre and @justinhorvitz think! I'm inclined to accept this pull request for precisely the reasons you mention: it fixes the interface of Bazel and leaves us free to refactor the internals, should we end up deciding that that's the right thing to do but I don't want to decide all by myself.

I am swayed by the argument that, even if we intend to eventually switch to the "proper" approach, this PR is a good step in that direction, because it makes all of the contact points with the external world clear in the code (nevermind the fact that it also fixes long-standing bugs). I say we move forward with this PR; I volunteer to co-review it.

coeuvre commented:
+1 to what @lberki and @tjgq already said.

tjgq commented on src/test/java/com/google/devtools/build/lib/vfs/FileSystemTest.java:
```diff
@@ -154,52 +155,57 @@ protected void expectNotFound(Path path) throws IOException {
+      try (var entriesStream = Files.list(directoryToRemove)) {
+        entries = entriesStream.toArray(java.nio.file.Path[]::new);
+      }
+    } catch (NotDirectoryException ignored) {
+    }
```
Can we let the exception propagate? (I wouldn't expect to see any tests calling this method on a non-directory...)

tjgq commented on src/test/java/com/google/devtools/build/lib/vfs/FileSystemTest.java:
```diff
@@ -154,52 +155,57 @@ protected void expectNotFound(Path path) throws IOException {
-          }
-          if (!currentFile.delete()) {
-            throw new IOException("Failed to delete '" + currentFile + "'");
+            Files.setPosixFilePermissions(
+                entry,
+                Sets.union(
+                    Files.getPosixFilePermissions(entry),
+                    ImmutableSet.of(PosixFilePermission.OWNER_WRITE)));
```
I don't understand this: on Unix you don't need write permission on a file to delete it, only on the parent directory. (I think it must be a Windows thing?)

tjgq commented on src/test/cpp/blaze_util_windows_test.cc:
```diff
@@ -128,7 +128,7 @@ TEST(BlazeUtilWindowsTest, TestSetEnv) {
   ASSERT_ENVVAR("Bazel_TEST_Key1", "some_VALUE");
   SetEnv("Bazel_TEST_Key1", "");
-  ASSERT_ENVVAR_UNSET("Bazel_TEST_Key1");
+  ASSERT_ENVVAR("Bazel_TEST_Key1", "");
```
I'm confused; how come this test was passing before?

tjgq replied:
Thanks for the explanation.

justinhorvitz commented:
> > Yeah, I agree that if we go with the "use proper UTF-8" approach, Googlers are in a better position to do so. In fact, there is a change from a few years ago floating about that just did that; I don't remember what its fate was.
> > Let's see what @tjgq and @coeuvre and @justinhorvitz think! I'm inclined to accept this pull request for precisely the reasons you mention: it fixes the interface of Bazel and leaves us free to refactor the internals, should we end up deciding that that's the right thing to do but I don't want to decide all by myself.
> 
> I am swayed by the argument that, even if we intend to eventually switch to the "proper" approach, this PR is a good step in that direction, because it makes all of the contact points with the external world clear in the code (nevermind the fact that it also fixes long-standing bugs). I say we move forward with this PR; I volunteer to co-review it.

I don't want to be a full reviewer of this PR, but I want to make the reviewers aware of a couple things with the way we handle interfacing with other internal systems, so you can make sure that's not broken. Please ping me and I'll give you the references.

justinhorvitz commented:
> > > Yeah, I agree that if we go with the "use proper UTF-8" approach, Googlers are in a better position to do so. In fact, there is a change from a few years ago floating about that just did that; I don't remember what its fate was.
> > > Let's see what @tjgq and @coeuvre and @justinhorvitz think! I'm inclined to accept this pull request for precisely the reasons you mention: it fixes the interface of Bazel and leaves us free to refactor the internals, should we end up deciding that that's the right thing to do but I don't want to decide all by myself.
> > 
> > 
> > I am swayed by the argument that, even if we intend to eventually switch to the "proper" approach, this PR is a good step in that direction, because it makes all of the contact points with the external world clear in the code (nevermind the fact that it also fixes long-standing bugs). I say we move forward with this PR; I volunteer to co-review it.
> 
> I don't want to be a full reviewer of this PR, but I want to make the reviewers aware of a couple things with the way we handle interfacing with other internal systems, so you can make sure that's not broken. Please ping me and I'll give you the references.

Oh, one of the things I was thinking of is actually open source: https://cs.opensource.google/bazel/bazel/+/master:src/main/native/latin1_jni_path.cc.

fmeum commented on src/main/java/com/google/devtools/build/lib/vfs/AbstractFileSystem.java:
```diff
@@ -100,18 +99,16 @@ private InputStream createMaybeProfiledInputStream(PathFragment path) throws IOE
     } finally {
       if (shouldProfile) {
-        profiler.logSimpleTask(startTime, ProfilerTask.VFS_OPEN, name);
+        profiler.logSimpleTask(startTime, ProfilerTask.VFS_OPEN, path.toString());
```
The writes ultimately end up here: https://github.com/bazelbuild/bazel/blob/424ff69579497768206bea1ae750b4add60262f3/src/main/java/com/google/devtools/build/lib/profiler/JsonTraceFileWriter.java#L227
It seems easier and more consistent to just have this `OutputStreamWriter` use Latin-1. What do you think? I would prefer to defer the changes that only affect "logging" in the broadest sense to a later PR.

tjgq replied:
I like that idea! It's much less error-prone, because in reality most Bazel developers (myself included) will forget to reencode error messages containing filenames some of the time.

Agreed wrt deferring all logging-related issues to a later PR.

lberki commented:
Review process: reviewed e54c6fbd39a661a5a0e68ed00fea39120ae1c91a , no comments other than those of @tjgq 's .

lberki commented on src/main/native/darwin/fsevents.cc:
```diff
@@ -160,11 +162,15 @@ Java_com_google_devtools_build_lib_skyframe_MacOSXFsEventsDiffAwareness_poll(
-    jclass classString = env->FindClass("java/lang/String");
-    result = env->NewObjectArray(info->paths.size(), classString, nullptr);
+    jclass classByteArray = env->FindClass("[B");
+    result = env->NewObjectArray(info->paths.size(), classByteArray, nullptr);
```
Is this how you create an Object array containing byte arrays? It's been ages that I wrote JNI code, but naively I'd think that you'd need to put `java.lang.Object` here since the return value is `Object[]` and not `byte[][]`.

lberki commented:
9eb4e1d83e1f9935f28569afcc167d7ef99689cc reviewed.

lberki commented on src/main/java/com/google/devtools/build/lib/util/StringEncoding.java:
```diff
@@ -0,0 +1,183 @@
+ *
+ * <p>While allowing the user a great deal of flexibility, this requires great care when {@link
+ * String}s are passed into or out of Bazel via Java standard library functions or external APIs.
+ * The following three difference types of strings need to be distinguished as if they were
```
nit: three *different* types of strings

lberki commented on src/main/java/com/google/devtools/build/lib/util/StringEncoding.java:
```diff
@@ -0,0 +1,183 @@
+    if (STRING_UNSAFE.getCoder(s) != StringUnsafe.LATIN1) {
+      throw new IllegalArgumentException(
+          "Expected internal string LATIN1 coded, got UTF16 bytes: %s (%s)"
+              .formatted(Arrays.toString(STRING_UNSAFE.getByteArray(s)), s));
```
nit: mind imposing an upper limit (dunno, 1000 chars?) on what's printed. Otherwise, it's potentially a lot of log spam.

lberki commented on src/main/java/com/google/devtools/build/lib/util/StringEncoding.java:
```diff
@@ -0,0 +1,183 @@
+ *       array {@link String#value} and the {@link String#coder} is {@link String#LATIN1}.
+ *   <li>Unicode strings: Regular Java strings, which are always Unicode. A common example is a
+ *       {@code string} field in a protobuf message.
+ *   <li>Platform strings: Strings that are passed to or returned from Java (N)IO functions or as
```
How are platform strings different from Unicode strings?

AFAIU Java Strings don't have a user-visible "encoding" value so I maybe you mean the encoding of their internal byte array? But that can't be cause you just use `new String()` in `internalToPlatform()` below just like in `internalToUnicode()`.

The distinction makes sense philosophically, but I thought the practical difference would only matter when transforming these String objects into byte arrays directly before calling the APIs of the operating system.

fmeum replied:
The only difference is the condition used to decide whether to reencode: platform strings may be Latin-1 arrays of raw bytes that we want to preserve as such, Unicode strings aren't.

lberki commented:
4a6e219850e135a5cc67760559283822644d4559 reviewed.

lberki commented:
Okay, I'm done with the review. That's was a LOT.

tjgq commented on src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java:
Same as above: I agree that using Latin1 for output streams is a superior approach (and also that cosmetic issues don't need to be addressed in this PR).

fmeum commented:
There is a persistent failure in the `watchfs_test` on macOS that I can't reproduce locally. I am pushing new commits to debug it in CI, but I won't change https://github.com/bazelbuild/bazel/pull/24010/commits/6373646b933a86adb2f3336a05be730f8b0c293d or any commits below it.

lberki commented:
To get an early signal on how difficult it would be to update Google-internal code after merging this pull request, I ran our presubmit battery and it ended up pleasantly green: the only failures are `UnixTests`, `VfsTests`, `run_test`, `watchfs_test` and an internal test that uses `FileSystem` creatively.

@justinhorvitz expressed some concerns about not catching all the places in Google-internal code where we currently re-encode strings in a way that's incompatible with this change. IMHO looking for references to `StandardCharsets.{ISO_8859_1,UTF_8} should be enough to flag these places and hopefully all one needs is to replace them with a call to the appropriate method on `StringEncoding`.

fmeum commented:
I still don't know why `watchfs_test` started to fail in CI (it does even if I revert my recent changes). I'll keep investigating this.

The other tests all set some specific locales via `env` or at runtime. It's possible that these locales aren't present in your test environment or that the Java stub template in Blaze overrides them.

fmeum commented:
Looks like explicitly setting a locale fixed the test, CI is green now.

fmeum commented:
I had to make `watchfs_test` exclusive and bump the sleeps to 5s to get it to pass on the Mac runners. It passes on my Mac with sleeps at 1s in 100/100 runs, so this seems to be specific to the runners.

tjgq commented on src/main/java/com/google/devtools/build/lib/util/StringEncoding.java:
Something to ponder: should we add a static initialization block in this class that tests for `COMPACT_STRINGS` and crashes the Bazel process otherwise? This might save some debugging if someone tries to run the Bazel server with compact strings disabled, or with a JVM that doesn't implement them (i.e., it won't look like a legit bug when the `coder == LATIN1` assertion later fails).

tjgq commented on src/main/native/latin1_jni_path.cc:
```diff
@@ -39,53 +39,24 @@ jstring NewStringLatin1(JNIEnv *env, const char *str) {
-  if (str == nullptr) {
+  // All path strings used in Bazel are encoded as raw bytes with a Latin1
+  // coder.
+  if (env->GetByteField(jstr, String_coder_field) != 0) {
```
`BAZEL_LOG(FATAL)` seems better than returning null?

tjgq commented on src/main/java/com/google/devtools/build/lib/skyframe/serialization/strings/UnsafeStringCodec.java:
```diff
@@ -38,7 +38,7 @@ public final class UnsafeStringCodec extends LeafObjectCodec<String> {
   private static final UnsafeStringCodec INSTANCE = new UnsafeStringCodec();
 
-  private final StringUnsafe stringUnsafe = StringUnsafe.getInstance();
+  private final StringUnsafe STRING_UNSAFE = StringUnsafe.getInstance();
```
`static`?

tjgq commented on src/main/java/com/google/devtools/build/lib/actions/ParameterFile.java:
```diff
@@ -81,119 +70,39 @@ public enum ParameterFileType {
-  private static void writeContentLatin1(OutputStream outputStream, Iterable<String> arguments)
+  private static void writeContent(OutputStream out, Iterable<String> arguments)
       throws IOException {
     StringUnsafe stringUnsafe = StringUnsafe.getInstance();
```
Nit - mind moving this into a `private static final`?

tjgq commented:
> I had to make `watchfs_test` exclusive and bump the sleeps to 5s to get it to pass on the Mac runners. It passes on my Mac with sleeps at 1s in 100/100 runs, so this seems to be specific to the runners.

cc @fweikert we should probably look into this.

tjgq reviewed (changes requested):
Just a couple more comments, but I'm generally happy with the current state.

lberki commented:
I did a bout of debugging to see how to fix the newly added tests, and I still don't know :(

I looked at `UnixDigestHashAttributeNameTest.testGetNioPath_externalUtf8[SHA-256].`

I do know how it breaks, though: our test battery cleverly unsets `LANG`, which means that `sun.jnu.encoding` will be `ANSI_X3.4-1968`, which is basically ASCII. Then `StringEncoding.internalToPlatform()` will do reencoding, and recover the original string with the basil/Japanese characters. Then, `sun.nio.fs.UnixPath.encode()` correctly determines that that cannot be represented if the platform encoding is ASCII and thus dies.

There are two issues:
1. Bazel hard-wires ISO8859-1 as LANG in its server binary, but our internal test battery uses an unset `LANG`, therefore we are testing things in a way that's not consistent with production. 
2. The test does not actually work with ISO8859-1 as `sun.jnu.encoding` because then it runs into the same "unmappable character" issue

The obvious solution would be to temporarily set `sun.jnu.encoding` in that test, but that doesn't seem to be possible: it doesn't look like the JVM honors `-Dsun.jnu.encoding` on the command line and setting it using `System.setProperty()` doesn't work because its value is cached in many places (e.g. `sun.nio.fs.Util.jnuEncoding`)

So other than reaching into that class by reflection, I don't see how the test can work other than by assuming that UTF-8 is the encoding it's running under, which is not the case at Google.

fmeum commented:
> The test does not actually work with ISO8859-1 as sun.jnu.encoding because then it runs into the same "unmappable character" issue

Did you verify this? I think it's not true: If `sun.jnu.encoding` is Latin-1 on Unix, where native filesystem APIs are byte-based, the JVM will behave just like `UnixFileSystem` and read and produce "raw bytes in Latin-1 coder" strings, which is good enough for the test to pass.

> Bazel hard-wires ISO8859-1 as LANG in its server binary, but our internal test battery uses an unset LANG, therefore we are testing things in a way that's not consistent with production.

The Bazel test uses Bazel's Java stub template, which always forces a UTF-8 locale for tests to run in. I don't know what Blaze's Java stub template looks like, but would it work to explicitly set `LANG` or `LC_ALL` via `env`, just like `StringEncodingTest` does?

lberki commented:
```
$ cat X.java
public class X {
  public static void main(String[] args) throws Exception {
    System.err.println(System.getProperty("sun.jnu.encoding"));
    java.nio.file.Paths.get(new String("🌱"));
  }
}
$ javac X.java
$ LANG=en_US.ISO8859-1 java X
ISO-8859-1
Exception in thread "main" java.nio.file.InvalidPathException: Malformed input or input contains unmappable characters: ?
        at java.base/sun.nio.fs.UnixPath.encode(UnixPath.java:131)
        at java.base/sun.nio.fs.UnixPath.<init>(UnixPath.java:77)
        at java.base/sun.nio.fs.UnixFileSystem.getPath(UnixFileSystem.java:312)
        at java.base/java.nio.file.Path.of(Path.java:148)
        at java.base/java.nio.file.Paths.get(Paths.java:69)
        at X.main(X.java:5)
```

So that seems to be the case.

tjgq commented:
> on Linux:
> 
> ```
> $ cat X.java
> 
> public class X {
>   public static void main(String[] args) throws Exception {
>     System.err.println(System.getProperty("sun.jnu.encoding"));
>     java.nio.file.Paths.get(new String("🌱"));
>   }
> }
> $ javac X.java
> $ LANG=en_US.ISO8859-1 java X
> ISO-8859-1
> Exception in thread "main" java.nio.file.InvalidPathException: Malformed input or input contains unmappable characters: ?
>         at java.base/sun.nio.fs.UnixPath.encode(UnixPath.java:131)
>         at java.base/sun.nio.fs.UnixPath.<init>(UnixPath.java:77)
>         at java.base/sun.nio.fs.UnixFileSystem.getPath(UnixFileSystem.java:312)
>         at java.base/java.nio.file.Path.of(Path.java:148)
>         at java.base/java.nio.file.Paths.get(Paths.java:69)
>         at X.main(X.java:5)
> ```
> 
> So that seems to be the case.

But the test isn't calling `Paths.get(new String("🌱"))`, it's calling `Paths.get(unicodeToInternal(new String("🌱")))` which should be equivalent to `Paths.get(new String("\xf0\x9f\x8c\xb1"))` (UTF-8 encoding of "🌱"), whose characters are all <= 0xff and thus decodable as ISO-8859-1?

lberki commented:
> But the test isn't calling `Paths.get(new String("🌱"))`, it's calling `Paths.get(unicodeToInternal(new String("🌱")))` which should be equivalent to `Paths.get(new String("\xf0\x9f\x8c\xb1"))` (UTF-8 encoding of "🌱"), whose characters are all <= 0xff and thus decodable as ISO-8859-1?

I see. I thought you meant that when `sun.jnu.encoding` is ISO8859-1, 🌱 can be encoded for some reason.

I thought about this and I think the root cause of the issue is that Bazel does one thing (UTF-8), Blaze does another thing (ISO8859-1) and the tests do a third thing (ASCII). I feel uneasy that the tests replicate the production environment that unfaithfully, but it's not an easy fix because there isn't a way to set `LANG` for Java tests in all scenarios (we unmodifiably unset `LANG` internally regardless of what `java_test.env` says so the trick you suggest doesn't work :(

With this in mind, I suggest deleting these tests and relying entirely on integration tests instead.  It's not great, but the second best alternative (wrapping the Java test binaries in `sh_test` and tweaking `LANG` that way) seems to be more pain than it's worth.

fmeum commented:
> I thought about this and I think the root cause of the issue is that Bazel does one thing (UTF-8), Blaze does another thing (ISO8859-1) and the tests do a third thing (ASCII).

Bazel and Blaze do the same thing here on Linux, it's just that most users won't have `en_US.ISO-8859-1` installed on their system. 

>  I feel uneasy that the tests replicate the production environment that unfaithfully, but it's not an easy fix because there isn't a way to set LANG for Java tests in all scenarios (we unmodifiably unset LANG internally regardless of what java_test.env says so the trick you suggest doesn't work :(

If you only unset `LANG`, could we use `LC_ALL` or `LC_CTYPE` instead?

> With this in mind, I suggest deleting these tests and relying entirely on integration tests instead. It's not great, but the second best alternative (wrapping the Java test binaries in sh_test and tweaking LANG that way) seems to be more pain than it's worth.

The tests would still be useful in Bazel's presubmit and it's far easier to provide comprehensive coverage of all file systems via a Java test. I would thus prefer to either wrap them in this way or to disable them for Blaze (e.g. via `assume` on the locale).

lberki commented:
Sigh, some documentation may be out of date. Contrary to what our internal documentation says, setting `LC_ALL` does actually work so we do have a reasonable path forward.

When setting that, there are two breakages:
* `testGetNioPath_internalUtf8` is rightfully broken because it tries to create a file whose name has a Unicode character, which obviously doesn't work if the encoding for path is ISO8859-1.
* `testUtf8Symlink` seems to have a genuine bug ("expected: 入力_A_🌱.target but was : å¥å_A_ð±.target" in the last line of the test. Mind taking a look? You can probably reproduce it by the above `env = {"LC_ALL": "en_US.ISO-8859-1"}` trick.

fmeum commented:
@lberki Thanks for reporting the breakages, I was able to reproduce them after forcing a Latin-1 locale on Linux. As it turns out, the two tests were simply missing conversions between Unicode and platform strings. They should be passing now.

fmeum commented:
The RBE environment doesn't have a Latin-1 locale installed. I am skipping the tests in that case, but maybe we could add it?

lberki commented:
@tjgq, are you fine with the current state of this pull request?

If so, I'll take a look at the internal references to `ISO_8859_1` and `UTF_8` to see if there are any obvious landmines in the code. The internal test battery looks fine -- aside from the aforementioned breakages that have all been fixed since then, everything seems to be green.

tjgq commented:
Yes, I am happy with the current state.

On Mon, Nov 4, 2024 at 13:06 lberki ***@***.***> wrote:

> @tjgq <https://github.com/tjgq>, are you fine with the current state of
> this pull request?
>
> If so, I'll take a look at the internal references to ISO_8859_1 and UTF_8
> to see if there are any obvious landmines in the code. The internal test
> battery looks fine -- aside from the aforementioned breakages that have all
> been fixed since then, everything seems to be green.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/24010#issuecomment-2454541915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBK5HOICVK5USXCLRZ2DA3Z65PKZAVCNFSM6AAAAABQB7H2RKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINJUGU2DCOJRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lberki commented:
Ack, I'll put it on my TODO list then (I'm on duty this week, though, so my bandwidth is limited, so no guarantees as to when I get around doing this)

fmeum commented:
@tjgq Note that these call sites don't convert `String`s to `String`s, they turn `String`s into `ByteString`s. That never requires reencoding, it just means "give me the raw bytes of the internal encoding". 

I actually had a helper function for this in `StringEncoding`, but it requires a dep on protobuf that is problematic for some consumers of `StringEncoding` (I recall some tricky issues with `bootstrap_java_library`). That's why I ended up removing it. 

We could add an `INTERNAL_CHARSET` constant in `StringEncoding` that is set to `ISO_8859_1` and may be flipped to `UTF_8` in the future (but not necessarily, we may very well choose to continue to represent strings as UTF-8 in memory). What do you think?

fmeum commented:
@bazel-io fork 8.0.0

tjgq commented:
> @tjgq Note that these call sites don't convert `String`s to `String`s, they turn `String`s into `ByteString`s. That never requires reencoding, it just means "give me the raw bytes of the internal encoding".
> 
> I actually had a helper function for this in `StringEncoding`, but it requires a dep on protobuf that is problematic for some consumers of `StringEncoding` (I recall some tricky issues with `bootstrap_java_library`). That's why I ended up removing it.
> 
> We could add an `INTERNAL_CHARSET` constant in `StringEncoding` that is set to `ISO_8859_1` and may be flipped to `UTF_8` in the future (but not necessarily, we may very well choose to continue to represent strings as UTF-8 in memory). What do you think?

I was hoping for a helper similar `STRING_UNSAFE.getInternalStringBytes` which, in addition to being self-documenting, would also give us an additional opportunity to catch bugs. But if that creates a problematic dependency, I say let's leave it for a separate PR; I don't want to delay this further, since we still have an internal review to do and I'd very much like it to make it into the 8.0.0 release.

tjgq commented:
@lberki lmk if you'd rather have me do the import instead

lberki commented:
@tjgq I'll give it a stab now, then if I fail, I'll hand over the baton

lberki commented:
Update: I imported the change and somewhat surprisingly, `remote_execution_test.test_unicode_execution` started failing, even though I only needed very minimal changes:

https://buildkite.com/bazel/google-bazel-presubmit/builds/85996#01930156-97b2-41cc-a133-8e331e07dd42

@tjgq , mind taking a look?

tjgq commented:
> Update: I imported the change and somewhat surprisingly, `remote_execution_test.test_unicode_execution` started failing, even though I only needed very minimal changes:
> 
> https://buildkite.com/bazel/google-bazel-presubmit/builds/85996#01930156-97b2-41cc-a133-8e331e07dd42
> 
> @tjgq , mind taking a look?

This was due to a bad import that was missing the changes in src/tools/remote/BUILD.

tjgq commented:
I'm going to replace `private static final Charset SUN_JNU_ENCODING` in `StringEncoding.java` with `private static final boolean SUN_JNU_ENCODING_IS_ISO_8859_1` because (1) reference equality isn't guaranteed to work for `Charset` and (2) we might as well inline the comparison, anyway.

fmeum commented:
> I'm going to replace `private static final Charset SUN_JNU_ENCODING` in `StringEncoding.java` with `private static final boolean SUN_JNU_ENCODING_IS_ISO_8859_1` because (1) reference equality isn't guaranteed to work for `Charset` and (2) we might as well cache the comparison, anyway.

Sounds good, this avoids assumptions on what the JIT may inline.

lberki commented:
@fmeum thanks for this pull request and your persistence. It really is a monster-sized one that makes Bazel better in ways that I didn't think was possible incrementally.

fmeum commented:
> @fmeum thanks for this pull request and your persistence. It really is a monster-sized one that makes Bazel better in ways that I didn't think was possible incrementally.

Thanks, it took me multiple failed attempts before this one that seems to work. I have two follow-up PRs in the pipeline, but they won't be monster-sized. :⁠-⁠) Thanks for the thorough reviews!
